packages/vulcan-lib/lib/modules/reactive-state.js

```diff
@@ -0,0 +1,215 @@
+ */
+export const setReactiveStateValue = (stateKey, value) => {
+  const stateObject = getReactiveState(stateKey);
+  value = cleanReactiveStateValue(value || stateObject.defaultValue, stateObject.schema);
+
+  stateObject.reactiveVar(value);
+  return stateObject.reactiveVar();
+};
+
+
+/**
+ * Updates the reactive state using immutability-helper
+ * @see {@link https://github.com/kolodny/immutability-helper)
+ *
+ * @example Add an element to an array field
+ *  updateReactiveStateValue('StarWars', { characters: { $push: ['Mayfeld'] } });
+ *
+ * @param {string} stateKey The key of the desired reactive state
+ * @param {Object} updates The updates to apply to the current state
+ * @return {Object} The new value of the reactive state
+ */
+export const updateReactiveStateValue = (stateKey, updates) => {
+  const stateObject = getReactiveState(stateKey);
+  let value = stateObject.reactiveVar();
+
+  if (typeof value === 'object' && typeof updates === 'object') {
+    value = update(value, updates);
+  }
+
+  return setReactiveStateValue(stateKey, value);
+};
+
+
+/**
+ * Resets the value of all reactive states to their defaults
+ */
+export const resetReactiveState = () => {
+  _forOwn(reactiveStates, function (stateObject, stateKey) {
+    setReactiveStateValue(stateKey);
+  });
+};
+
+
+/**
+ * React hook for using reactive state in a functional component
+ * @param {Object} options - Hook options
+ *        {string} options.stateKey - The key of the desired reactive state
+ *        {string} options.[propName] - Name of the props injected by the hook; omit to use `stateKey`; this is
+ *        useful if your code needs to work with various states without knowing their keys
+ * @returns {Object} props - Props returned by the hook; the name of the props will be based on `stateKey` or
+ * `propName` (`StateName` is just an example)
+ * @returns {ReactiveState} props.StateName - The reactive state
+ * @returns {function} props.setStateName - A function to set the state
+ * @returns {function} props.updateStateName - A function to update the state
+ */
+export const useReactiveState = (options) => {
+  const {stateKey, propName = stateKey} = options;
+  const stateObject = getReactiveState(stateKey);
+
+  const props = {
+    [propName]: stateObject,
+    [`set${propName}`]: function (value) {
+      return setReactiveStateValue(stateKey, value);
+    },
+    [`update${propName}`]: function (updates) {
+      return updateReactiveStateValue(stateKey, updates);
+    },
+  };
+
+  return props;
+};
+
+
+/**
+ * React HoC for using reactive state in a Class component
+ * @param {Object} options Hook options
+ *        {string} options.stateKey The key of the desired reactive state
+ *        {string} options.[propName] Name of the props injected by the hook; omit to use `stateKey`
+ * @returns {function} A wrapped React component that includes the props returned by `useReactiveState`
+ */
+export const withReactiveState = (options) => C => {
```

Comment:
Do you use the HOC in your own app? Otherwise it can quickly become more maintenance, I'd expect peoeple to write their own HOC if they really need them.

Reply:
I do use HoC's primarily because most of my code was written before hooks existed. It's my opinion that for now we should provide both a hook and an HoC wrapper for all core functionality.